src/Symfony/Component/HttpClient/NativeHttpClient.php

```diff
@@ -0,0 +1,400 @@
+            }
+        };
+
+        // Resolve the host's IP using the local DNS cache if possible
+        if ($port = parse_url($url['authority'], PHP_URL_PORT) ?: '') {
+            $info['primary_port'] = $port;
+            $port = ':'.$port;
+        }
+        $host = parse_url($url['authority'], PHP_URL_HOST);
+
+        if ($options['resolve']) {
+            $this->multi->dnsCache = $options['resolve'] + $this->multi->dnsCache;
+        }
+
+        if (null === $ip = $this->multi->dnsCache[$host] ?? null) {
+            $now = microtime(true);
+            if (!$ip = gethostbynamel($host)) {
+                throw new TransportException(sprintf('Could not resolve host "%s".', $host));
+            }
+            $info['namelookup_time'] = microtime(true) - $now;
+            $this->multi->dnsCache[$host] = $ip = $ip[0];
+        }
+
+        $info['primary_ip'] = $ip;
+        $url['authority'] = substr_replace($url['authority'], $ip, -\strlen($host) - \strlen($port), \strlen($host));
+
+        if (!isset($options['headers']['host'])) {
+            $options['raw_headers'][] = 'host: '.$host.$port;
+        }
+
+        if ($onProgress) {
+            // Notify DNS resolution
+            $onProgress();
+        }
+
+        $context = [
+            'http' => [
+                'protocol_version' => $options['http_version'] ?: '1.1',
+                'method' => $method,
+                'content' => $options['body'],
+                'ignore_errors' => true,
+                'user_agent' => 'Symfony HttpClient/Native',
+                'curl_verify_ssl_peer' => $options['verify_peer'],
+                'curl_verify_ssl_host' => $options['verify_host'],
+                'auto_decode' => false, // Disable dechunk filter, it's incompatible with stream_select()
+                'timeout' => $options['timeout'],
+                'follow_location' => false, // We follow redirects ourselves - the native logic is too limited
+            ],
+            'ssl' => array_filter([
+                'peer_name' => $host,
+                'verify_peer' => $options['verify_peer'],
+                'verify_peer_name' => $options['verify_host'],
+                'cafile' => $options['cafile'],
+                'capath' => $options['capath'],
+                'local_cert' => $options['local_cert'],
+                'local_pk' => $options['local_pk'],
+                'passphrase' => $options['passphrase'],
+                'ciphers' => $options['ciphers'],
+                'peer_fingerprint' => $options['peer_fingerprint'],
+                'capture_peer_cert_chain' => $options['capture_peer_cert_chain'],
+                'allow_self_signed' => (bool) $options['peer_fingerprint'],
+                'SNI_enabled' => true,
+                'disable_compression' => true,
+            ], static function ($v) { return null !== $v; }),
+            'socket' => [
+                'bindto' => $options['bindto'],
+                'tcp_nodelay' => true,
```

Comment:
IMO this option should be set to false, true is very useful on small request, but on medium to large stream (like http) we want to avoid congestion by default.

Reply:
Both fopen and curl already buffer as much as they can before writing to the network. This means TCP nodelay is not needed and only degrades performance.